flex_model/core/wrapper.py

```diff
@@ -59,13 +61,170 @@ class _SharedState:
+
+        # Associate the hook functions with this group.
+        for hook_fn in hook_functions_to_add:
+            self.hook_fn_to_groups_map[hook_fn].add(group_name)
+
+        if group_name not in self.groups:
+            self.groups.add(group_name)
+
+    def _is_group_alive(self, group_name: str) -> bool:
+        """A group is alive if it has one or more associated hook functions."""
+        return any(
+            map(lambda gs: group_name in gs, list(self.hook_fn_to_groups_map.values()))
+        )
+
+    @functools.singledispatchmethod
+    def remove(self, group_constructor, group_name) -> None:
+        raise NotImplementedError
+
+    @remove.register
+    def _remove_by_list_hook_fns(
+        self, group_constructor: list, group_name: str
+    ) -> None:
+        are_hook_fns = map(lambda x: isinstance(x, HookFunction), group_constructor)
+        assert all(are_hook_fns), f"set_group takes a collection of only HookFunctions"
+        assert group_name != "all", "Can't remove 'all' group references"
+
+        for hook_fn in group_constructor:
+            self.hook_fn_to_groups_map[hook_fn].remove(group_name)
+
+        if not self._is_group_alive(group_name):
+            self.groups.remove(group_name)
+
+    @remove.register
+    def _remove_by_hook_fn(
+        self, group_constructor: HookFunction, group_name: str
+    ) -> None:
+        self.hook_fn_to_groups_map[group_constructor].remove(group_name)
+
+        if not self._is_group_alive(group_name):
+            self.groups.remove(group_name)
+
+    @remove.register
+    def _remove_by_string(self, group_constructor: str, group_name: str) -> None:
+        hook_functions_to_remove = []
+        for hook_fn, groups in self.hook_fn_to_groups_map.items():
+            if group_constructor in hook_fn.module_name:
+                hook_functions_to_remove.append(hook_fn)
+
+        for hook_fn in hook_functions_to_remove:
+            self.hook_fn_to_groups_map[hook_fn].remove(group_name)
+
+        if not self._is_group_alive(group_name):
+            self.groups.remove(group_name)
+
+    def bisect(self, active_group_names: Union[str, List[str]]) -> Set[HookFunction]:
+        """Separate all hook functions into active/inactive sets."""
+        if isinstance(active_group_names, str):
+            active_group_names = [active_group_names]
+        active_group_names = set(active_group_names)
+
+        # A hook function is active if its intersection with the requested
+        # active groups is nonzero.
+        active_hook_fns = set()
+        inactive_hook_fns = set()
+        for hook_fn, groups in self.hook_fn_to_groups_map.items():
+            if not groups.isdisjoint(active_group_names):
+                active_hook_fns.add(hook_fn)
+            else:
+                inactive_hook_fns.add(hook_fn)
+
+        # Sanity check: Hook functions belong to only one group.
+        assert active_hook_fns.isdisjoint(inactive_hook_fns)
```

Comment:
I think for public API we should raise exceptions instead of assertions (I believe this is also convention - see https://google.github.io/styleguide/pyguide.html#244-decision).

Reply:
This is a private class which the user is not expected to use. It should be fine to have an assertion here since we are asserting internal correctness (ie. disjoint invariant of our active/inactive group sets).